src/resolution/snapshot.rs

```diff
@@ -0,0 +1,601 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::hash_map;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use anyhow::bail;
+use anyhow::Context;
+use anyhow::Error as AnyError;
+use deno_semver::npm::NpmPackageNv;
+use deno_semver::npm::NpmPackageReq;
+use deno_semver::VersionReq;
+use futures::future::Either;
+use futures::StreamExt;
+use log::debug;
+use serde::Deserialize;
+use serde::Serialize;
+
+use super::common::resolve_best_package_version_info;
+use super::common::LATEST_VERSION_REQ;
+use super::graph::Graph;
+use super::graph::GraphDependencyResolver;
+
+use crate::registry::NpmPackageInfo;
+use crate::registry::NpmPackageVersionDistInfo;
+use crate::registry::NpmRegistryApi;
+use crate::NpmPackageCacheFolderId;
+use crate::NpmPackageId;
+use crate::NpmResolutionPackage;
+
+/// Packages partitioned by if they are "copy" packages or not.
+pub struct NpmPackagesPartitioned {
+  pub packages: Vec<NpmResolutionPackage>,
+  /// Since peer dependency resolution occurs based on ancestors and ancestor
+  /// siblings, this may sometimes cause the same package (name and version)
+  /// to have different dependencies based on where it appears in the tree.
+  /// For these packages, we create a "copy package" or duplicate of the package
+  /// whose dependencies are that of where in the tree they've resolved to.
+  pub copy_packages: Vec<NpmResolutionPackage>,
+}
+
+impl NpmPackagesPartitioned {
+  pub fn into_all(self) -> Vec<NpmResolutionPackage> {
+    let mut packages = self.packages;
+    packages.extend(self.copy_packages);
+    packages
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct NpmResolutionSnapshotCreateOptionsPackage {
+  pub pkg_id: NpmPackageId,
+  pub dist: NpmPackageVersionDistInfo,
+  /// Key is what the package refers to the other package as,
+  /// which could be different from the package name.
+  pub dependencies: HashMap<String, NpmPackageId>,
+}
+
+#[derive(Debug, Clone)]
+pub struct NpmResolutionSnapshotCreateOptions {
```

Comment:
This struct was added to allow creating one of these from the lockfile. Please let me know if you have a better suggestion for the name.

Reply:
Name seems fine, but it would be helpful to have some docstrings explaining how each option field will impact the final snapshot